FlowCal/plot.py

```diff
@@ -379,20 +367,18 @@ def density2d(data,
     data_plot = data[:, channels]
 
     # If bins are not specified, try to get bins from data object
-    if (bins is None and hasattr(data_plot, 'hist_bin_edges')
-            and data_plot.hist_bin_edges(0) is not None
-            and data_plot.hist_bin_edges(1) is not None):
-        # Get bin information
-        bdx = data_plot.hist_bin_edges(0)
-        bdy = data_plot.hist_bin_edges(1)
-        # Get bin scaled indices
-        xdx = np.linspace(0, 1, len(bdx))
-        xsx = np.linspace(0, 1, (len(bdx) - 1)/div + 1)
-        xdy = np.linspace(0, 1, len(bdy))
-        xsy = np.linspace(0, 1, (len(bdy) - 1)/div + 1)
-        # Generate sub-sampled bins
-        bins = np.array([np.interp(xsx, xdx, bdx), 
-                            np.interp(xsy, xdy, bdy)])
+    if (hasattr(data_plot, 'hist_bins')
+            and data_plot.hist_bins(0) is not None
+            and data_plot.hist_bins(1) is not None):
+        # Duplicate bins into a two-element list if necessary
+        if not isinstance(bins, list):
+            bins = [bins, bins]
+        # X Axis
+        if bins[0] is None or isinstance(bins[0], int):
+            bins[0] = data_plot.hist_bins(0, nbins=bins[0], log=xlog)
+        # Y axis
+        if bins[1] is None or isinstance(bins[1], int):
+            bins[1] = data_plot.hist_bins(1, nbins=bins[1], log=ylog)
```

Comment:
I'm trying to replicate the behavior of `bins` here: http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.histogram2d.html

> * i don't understand this section. you were originally checking if `bins` is `None` (i.e. only extracting bins from data object if `bins` was not specified). Shouldn't this still be the desired behavior?

I believe what I'm doing now is at least equivalent, if not better. If `data_plot` does not have `hist_bins`, there is nowhere that we can generate `bins` from.

> * can you add the name of the first argument to `hist_bins()`? (as as you've done with the other 2 arguments) i forgot what that was and had to go look it up.

Will do.

> * was unclear what `data_plot.hist_bins(#)` was doing. still not 100% sure. why are you checking if that's none? when does that function return `None`?

`data_plot.hist_bins` can generate a requested number of bins using the `nbins` argument. If `nbins` is None, however, `data_plot.hist_bins` will use the resolution of the specified channel.

If the user specifies `bins` as an integer, `plot.hist1d` generates the requested number of bins using `data_plot.hist_bins`. Therefore, if `bins` is None or if it's an integer, it should be passed to `data_plot.hist_bins`.

> * description of `bins` also didn't make sense. Seems like you should only extract `bins` from data when `bins` is `None`

See the previous point.

> * does `isinstance(bins,list)` fail with tuples or array-like? should that just be checking for iterable?

No, actually if `bins` is an array I'm assuming that it specifies a set of bin edges.